Feature/QueryParams

gabriofr opened the pull request:
Vi fant ut at det ikke var hensiktsmessig å implementere react-router på nåværende tidspunkt, implementerte heller query-parameter for alle filter. 

Oppretter denne som draft, det er litt å se igjennom :)

Eksempel på bruk: http://localhost:3000/?chart_type=bar&level=M&selected_row=hjerneslag_beh_tromb&tu_names=Narvik_Harstad_Troms%C3%B8_Hammerfest_Lillehammer&year=2016

PS: Er ikke klar med alle testene enda.

LarsJK commented on src/App.tsx:
```diff
@@ -156,10 +156,35 @@ function APP() {
+  const get_valid_year = (year_to_validate: number, valid_years: number[]) => {
+    const max_year = Math.max.apply(null, opts_year);
+    const min_year = Math.min.apply(null, opts_year);
+    return year_to_validate > max_year
+      ? max_year
+      : year_to_validate < min_year
+      ? min_year
+      : year_to_validate;
```
I like this:

```suggestion
    return Math.max(min_year, Math.min(number, max_year));
```

Maybe define it as a global helper like this:

`const clamp = (input, min, max) => Math.max(min, Math.min(input, max));`

There is a proposal for Math.clamp but its not ready yet.

gabriofr replied:
Love it! <3 so much cleaner now. 
ps: eslint wouldn't let me extend Math.prototype so added a helper function.

LarsJK commented on src/App.tsx:
```diff
@@ -183,10 +208,12 @@ function APP() {
+  } = {
+    selected_unit: treatment_units as string[],
     selected_year: selected_year,
   };
```
You shouldn't need to specify types here?

```suggestion
  const input_data = {
    selected_unit: treatment_units,
    selected_year: selected_year,
  };
```

Is TS not able to infer the correct type like this?

gabriofr replied:
ah nice! it didn't at first, but it does now - there was som initial inference problems..

LarsJK commented on src/components/legend_btn.tsx:
```diff
@@ -16,14 +18,19 @@ function LEGEND_BTN(props: Props) {
     show_level_filter,
   } = props;
-
+  const [, set_level_query_param] = useQueryParam(
+    "level",
+    mainQueryParamsConfig.level
+  );
```
Will this not override the current query param on mount?

gabriofr replied:
nope, it just loads the setter with the config from mainQueryParamsConfig. 

```javascript
useQueryParam<T>(name: string, paramConfig: QueryParamConfig<T>, rawQuery?: ParsedQuery):
  [T | undefined, (newValue: T, updateType?: UrlUpdateType) => void]
```

LarsJK commented on src/index.tsx:
```diff
@@ -3,11 +3,20 @@ import React from "react";
+      <BrowserRouter>
+        <QueryParamProvider ReactRouterRoute={Route}>
+          <Route
+            path="/:indicator?/:level?/:year?/:tus?/:module?"
```
Should this be removed?

gabriofr replied:
Yes, leftovers from react-router-dom tests

LarsJK reviewed (changes requested):
I don't think it is necessary to do the useQueryParam > useState > useEffect loop.

We should be able to just swap out 
`const [value, setValue] = useState(initial);` 
with 
`const [value, setValue] = useQueryParam(key, initial);`

gabriofr commented on src/App.tsx:
```diff
@@ -156,10 +156,35 @@ function APP() {
-  const [selected_row, update_selected_row] = useState(null);
+  let opts_year = [2019, 2018, 2017, 2016];
+  const get_valid_year = (year_to_validate: number, valid_years: number[]) => {
+    const max_year = Math.max.apply(null, opts_year);
```
yup, it doesn't use this in this case, so I just passed inn whatever (null), but your suggestion is better. I will refaktor. :)

gabriofr replied:
fixed! also bug

gabriofr commented on src/App.tsx:
```diff
@@ -156,10 +156,35 @@ function APP() {
+  let opts_year = [2019, 2018, 2017, 2016];
+  const get_valid_year = (year_to_validate: number, valid_years: number[]) => {
+    const max_year = Math.max.apply(null, opts_year);
+    const min_year = Math.min.apply(null, opts_year);
```
Same

gabriofr replied:
Implemented your suggestions and found a bug at the same time, should be ...valid_years not ...opts_year... 
opts_year just happened to be declared above the function :p

gabriofr commented:
> I don't think it is necessary to do the useQueryParam > useState > useEffect loop.
> 
> We should be able to just swap out
> `const [value, setValue] = useState(initial);`
> with
> `const [value, setValue] = useQueryParam(key, initial);`

I've been thinking the same, it became obvious when I got all state covered with query-params.. 
I'll give it a shot :) 

PS: the useQueryParam hook is a bit weird, the second parameter is a QueryParamConfig, derived from a global config or set inline. Initial values are either set from defaults in the config or by assignment on the getter, or both.... 
`const [value=initial, setValue] = useQueryParam(key, QueryParamsConfig.key)`

gabriofr commented:
> I don't think it is necessary to do the useQueryParam > useState > useEffect loop.
> 
> We should be able to just swap out
> `const [value, setValue] = useState(initial);`
> with
> `const [value, setValue] = useQueryParam(key, initial);`

aight. removed the loop for the year param and it looks much better. Will continue with the rest :)

gabriofr commented:
Refactored a bit more. Removed all unnecessary state and effects. Added a validators where they would break the app and updated tests. Had to re-implement react-router-dom as useQueryParamProvider depends on it, and as we discussed - we will need it in the future. 

Also removed quite a few cases where state and setState was passed down, but unused. And moved QueryParams down to the child component where I could.

This should be ready now

LarsJK commented on src/components/TF_FIGURE/index.tsx:
```diff
@@ -25,7 +27,13 @@ function TF_FIGURE(props: Props) {
+    string | undefined
+  >("chart_type", mainQueryParamsConfig.chart_type);
+  const valid_chart_type = chart_type === "bar" ? "bar" : "line";
+  chart_type !== valid_chart_type &&
```
Same. Dont think we need to remove typos from url

gabriofr replied:
I concur

LarsJK commented on src/components/TF_FIGURE/index.tsx:
```diff
@@ -6,13 +6,15 @@ import TF_LONG_DESCRIPTION from "../tf_description";
   figure_class?: string;
   selectedTreatmentUnits: string[];
-  update_selected_row(row: string): void;
+  update_selected_row(row: string | undefined): void;
```
Do we have to change this to take undefined? Seems like nothing has changed in regards to this?

gabriofr replied:
undefined is (in useQueryParam) used to remove the param from the url. Don't think we need it now that I set a default and also validate. Type error is also gone now so I have removed it.

LarsJK commented on src/components/indicator_row.tsx:
```diff
@@ -131,7 +133,7 @@ function INDICATOR_ROW(props: Props) {
         {indicator_val}
         <INDICATOR_VALUE
-          key={``}
+          // key={``}
```
```suggestion
```

LarsJK commented on src/components/med_field_row.tsx:
```diff
@@ -0,0 +1,47 @@
+    update_med_field_filter(med_field.key);
+  };
+
+  useEffect(() => {
+    if (med_field.react_key === clicked_med_field) {
+      update_med_field_filter(med_field.key);
+    }
+  }, [med_field, clicked_med_field, update_med_field_filter]);
```
Do we need this? Looks like the click handler updates both.

gabriofr replied:
You are right, it's redundant.

LarsJK reviewed (changes requested):
Commented some small potential issues, but looks good otherwise 👍

gabriofr commented on src/App.tsx:
```diff
@@ -156,10 +181,31 @@ function APP() {
+    treatment_units as string[],
+    tu_names
+  );
+  JSON.stringify(treatment_units) !==
+    JSON.stringify(validated_treatment_units) &&
+    treatment_units &&
+    update_treatment_units(validated_treatment_units);
```
I see your point, this was done with intent as some props derived from query params actually break the app. It's ment more as a sanitiser than a type checker. Since we replaced state, query params are sent "blindly" down the line. We can remove the validators, but then we will have no sanitation what is delivered to  other components.

Also I'm not sure the use-case is to manipulating the url to achieve filtered results, I based it more on a point and click then share the url use-case. You can of course manipulate the url directly, but that will require "inside" knowledge of the params.

I suggest we leave it as is until we have refactored more of the app.

gabriofr replied:
Params are validated before they are handed down, I have removed the param type checker.. :)

gabriofr commented on src/App.tsx:
```diff
@@ -62,6 +64,30 @@ export interface AggData {
   all_filtered_by_year: StatisticData[];
 }
 
+function validate_year(year_to_validate: number, valid_years: number[]) {
```
Not sure about skipped years, can't think of a reason for supporting them. The `valid_years` array is imported from config as of now, I'll change it to a minYear + maxYear param. I agree it's easier and safer to maintain.

gabriofr commented on src/App.tsx:
```diff
@@ -156,10 +181,31 @@ function APP() {
+    selected_year || valid_years[0],
+    valid_years
+  );
+  validated_selected_year !== selected_year &&
+    selected_year &&
+    update_selected_year(validated_selected_year);
```
Not fixing it will break the single select box

gabriofr commented on src/App.tsx:
```diff
@@ -62,6 +64,30 @@ export interface AggData {
+      ?.filter((x, i, a) => {
+        return (
+          valid_treatment_units.some((tu) => tu.hospital === x) &&
+          a.indexOf(x) === i
```
Fixed

LarsJK commented on src/App.tsx:
```diff
@@ -62,6 +63,19 @@ export interface AggData {
+  return (
+    (treatment_units
+      ?.filter((x) =>
+        valid_treatment_units.some((tu) => tu.hospital === x || tu.hf === x)
```
What about rhf?

gabriofr replied:
RHFs are not selectable in the "show all screen", assumed this excludes them from query params. 

![image](https://user-images.githubusercontent.com/1658284/105717405-5edbfe00-5f20-11eb-9442-a19fa8ca502e.png)

arnfinn replied:
They are selectable from the search box:
![image](https://user-images.githubusercontent.com/136346/105718103-2c7ed080-5f21-11eb-8589-def1a2c9965e.png)

gabriofr replied:
Ah, then I assumed wrong - I'll add RHFs

gabriofr replied:
Done.
